core/src/main/java/org/infinispan/container/offheap/BoundedOffHeapDataContainer.java

```diff
@@ -209,13 +209,47 @@ private void ensureSize() {
                   addressToRemove = UNSAFE.getLong(firstAddress);
                }
             } else {
-               // This is the only way to break out of loop
+               // This and below are the only ways to break out of entire loop
                break;
             }
          } finally {
             lruLock.unlock();
          }
 
+         // If this was 0 that means we couldn't acquire the write lock with a try so we have to do a regular lock
+         if (addressToRemove == 0) {
+            // Acquire the write lock first and then double check the size before removing
+            entryWriteLock.lock();
+            try {
+               // Now we have to verify the size was greater than max again
+               lruLock.lock();
+               try {
+                  if (currentSize > maxSize) {
+                     // Retrieve the hashCode and make sure it has the same write lock owner
```

Comment:
Which hash code? And does the code below really check the owner?

Reply:
It is the hash code of the head element in the lru stack. We use that to find which lock to acquire.

Reply:
It is the hash code of the head element in the lru stack. We use that to find which lock to acquire.